packages/flutter/lib/src/foundation/change_notifier.dart

```diff
@@ -205,12 +221,13 @@ class ChangeNotifier implements Listenable {
   @visibleForTesting
   void notifyListeners() {
     assert(_debugAssertNotDisposed());
-    if (_listeners != null) {
-      final List<VoidCallback> localListeners = List<VoidCallback>.from(_listeners!);
-      for (final VoidCallback listener in localListeners) {
+    _firstNewlyAddedEntry = null;
+    if (_listeners != null && _listeners!.isNotEmpty) {
+      for (_ListenerEntry? entry = _listeners!.first;
+            entry != null && entry != _firstNewlyAddedEntry;
+            entry = entry.next) {
         try {
-          if (_listeners!.contains(listener))
-            listener();
+          entry.listener();
```

Comment:
What if `entry.listener()` calls `removeListener` for itself? Wouldn't that set `entry.next` to null and the iteration would end early?

Reply:
(If we don't have a test for this scenario, we should add one)

Reply:
Good catch

I should add tests for `addListener(() { notifier.notifyListeners() })` too